### PR TITLE
Add new cluster admin kubeconfig for aro service

### PR DESCRIPTION
**Pull request by olga-mir**
This is a re-make of https://github.com/Azure/ARO-RP/pull/250
It stores a new context for aro-service in a separate file as requested. 
It doesn't create additional role binding. 

I wasn't able yet to confirm that we can distinguish between the two identities in the audit logs. I've checked `/var/log/openshift-apiserver/audit.log` on all masters, but haven't found any action done by any of them (maybe we need to update audit policy to include these actions?)

Since we don't want to leverage kubeconfig context to manage context, we will need to either 
1. decide that `aro-service` kubeconfig is the admin kubeconfig we (as SRE team) are always using and ignore the default admin kubeconfig entirely.
2.  or manage multiple files ourselves and make sure KUBECONFIG points to the right place.

### CODE REVIEW

File: docs/deploy-development-rp.md [hunk 1]
```diff
@@ -131,7 +131,7 @@
 * Get an admin kubeconfig:
 
   ```bash
-  hack/get-admin-kubeconfig.sh "/subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/$RESOURCEGROUP/providers/Microsoft.RedHatOpenShift/openShiftClusters/$CLUSTER" >admin.kubeconfig
+  hack/get-admin-kubeconfig.sh "/subscriptions/$AZURE_SUBSCRIPTION_ID/resourceGroups/$RESOURCEGROUP/providers/Microsoft.RedHatOpenShift/openShiftClusters/$CLUSTER" admin >admin.kubeconfig
```
Comment (olga-mir) [inline docs/deploy-development-rp.md]: if we go with option 1 from the PR description we can remove addional param. For now I prefer to make it explicit. 
Better yet create Makefile targets (I noticed we don't have many helper targets at the moment)

File: pkg/install/kubeconfig.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	data, err := yaml.Marshal(aroServiceConfig)
+	if err != nil {
+		return err
+	}
+
+	var aroServiceInternalClient kubeconfig.AdminInternalClient
+	aroServiceInternalClient.File = &asset.File{
+		Filename: adminInternalClient.File.Filename, // need to change the name
```
Comment (olga-mir) [inline pkg/install/kubeconfig.go]: I will follow up on this.

**Comment by olga-mir** [comment]
CI failure:
```
ImportError: cannot import name timezone
Makefile:118: recipe for target 'test-python' failed
make: *** [test-python] Error 1
```
can't see how it is related to my changes. And passes locally

File: hack/get-admin-kubeconfig.sh [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
 
-go run ./hack/db "$1" | jq -r .openShiftCluster.properties.adminKubeconfig | base64 -d | sed -e 's|https://api-int\.|https://api\.|'
+if [[ "$CONTEXT_NAME" == "admin" ]]; then
+  KEY="adminKubeconfig"
+elif [[ "$CONTEXT_NAME" == "aro-service" ]]; then
+  KEY="aroServiceKubeconfig"
+else
+  echo "usage: context name must be one of admin or aro-service"
+  exit 1
+fi
```
Comment (jim-minter) [inline hack/get-admin-kubeconfig.sh]: why do we need this change, i.e. under what circumstances do we want to get the aro-service kubeconfig?

File: pkg/api/openshiftcluster.go [hunk 4]
```diff
@@ -85,9 +85,10 @@ type OpenShiftClusterProperties struct {
 
 	StorageSuffix string `json:"storageSuffix,omitempty"`
 
-	SSHKey            SecureBytes  `json:"sshKey,omitempty"`
-	AdminKubeconfig   SecureBytes  `json:"adminKubeconfig,omitempty"`
-	KubeadminPassword SecureString `json:"kubeadminPassword,omitempty"`
+	SSHKey               SecureBytes  `json:"sshKey,omitempty"`
+	AdminKubeconfig      SecureBytes  `json:"adminKubeconfig,omitempty"`
+	AroServiceKubeconfig SecureBytes  `json:"aroServiceKubeconfig,omitempty"`
```
Comment (jim-minter) [inline pkg/api/openshiftcluster.go]:
```suggestion
	AROServiceKubeconfig SecureBytes  `json:"aroServiceKubeconfig,omitempty"`
```

File: pkg/install/kubeconfig.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+// generateAndStoreKubeconfigs generates additional admin credentials and kubeconfig based admin kubeconfig
+// found in graph and stores both to the database
+func (i *Installer) generateAndStoreKubeconfigs(ctx context.Context, g graph, aroServiceName string) error {
+
+	ca := g[reflect.TypeOf(&tls.AdminKubeConfigSignerCertKey{})].(*tls.AdminKubeConfigSignerCertKey)
+	clientCertKey := g[reflect.TypeOf(&tls.AdminKubeConfigClientCertKey{})].(*tls.AdminKubeConfigClientCertKey)
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: I think you can just say `var clientCertKey tls.SignedCertKey` instead of this line - move to around line 38

File: pkg/install/kubeconfig.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+	cfg := &tls.CertCfg{
+		Subject:      pkix.Name{CommonName: aroServiceName, Organization: []string{"system:masters"}},
+		KeyUsages:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+		Validity:     tls.ValidityTenYears,
+	}
+
+	// new key and certificate will be stored in k.KeyRaw and k.CertRaw
+	err := clientCertKey.SignedCertKey.Generate(cfg, ca, "admin-kubeconfig-client", tls.DoNotAppendParent)
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: don't hard-code admin-kubeconfig-client, build a name from aroServiceName

File: pkg/install/kubeconfig.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	aroServiceInternalClient.Config = &aroServiceConfig
+
+	i.doc, err = i.db.PatchWithLease(ctx, i.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
+		// used for the SAS token with which the bootstrap node retrieves its
+		// ignition payload
+		doc.OpenShiftCluster.Properties.Install.Now = time.Now().UTC()
+		doc.OpenShiftCluster.Properties.AdminKubeconfig = adminInternalClient.File.Data
+		doc.OpenShiftCluster.Properties.AroServiceKubeconfig = aroServiceInternalClient.File.Data
+		return nil
+	})
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: leave this where it was, and have this function return the aro service kubeconfig to the caller.  It will make it easier to unit test.

File: pkg/install/kubeconfig.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+	// new key and certificate will be stored in k.KeyRaw and k.CertRaw
+	err := clientCertKey.SignedCertKey.Generate(cfg, ca, "admin-kubeconfig-client", tls.DoNotAppendParent)
+	if err != nil {
+		return err
+	}
+
+	// create a Config for the new service kubeconfig based on the generated cluster admin Config
+	clusters := make([]clientcmd.NamedCluster, len(adminInternalClient.Config.Clusters))
+	copy(clusters, adminInternalClient.Config.Clusters)
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: why is this needed? it looks like it's doing a shallow copy of a value which is not changed.

File: pkg/install/kubeconfig.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	if err != nil {
+		return err
+	}
+
+	var aroServiceInternalClient kubeconfig.AdminInternalClient
+	aroServiceInternalClient.File = &asset.File{
+		Filename: adminInternalClient.File.Filename, // need to change the name
+		Data:     data,
+	}
+	aroServiceInternalClient.Config = &aroServiceConfig
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: I don't think this is needed.  `data` is what you need, and you have it already

**Comment by olga-mir** [comment]
with the latest changes I was able to spot audit logs with different `usernames` eg.
```
"user":{"username":"system:aro-service","groups":["system:masters","system:authenticated"]},
"user":{"username":"system:admin","groups":["system:masters","system:authenticated"]}
```

**Comment by m1kola** [comment]
Do we need to update `pkg/util/restconfig/restconfig.go` to use the new config?

**Comment by olga-mir** [comment]
@m1kola yes, I had an unpushed commit, but you can see this file now here: https://github.com/Azure/ARO-RP/pull/262/files#diff-d3b408c550bc18f98048519469d20931

**Comment by olga-mir** [comment]
the unit test is still WIP it needs to have a few more checks, but the main scaffolding is done.

**Comment by olga-mir** [comment]
need to fix vet error:
```
pkg/install/kubeconfig_test.go:31:57: github.com/Azure/ARO-RP/vendor/github.com/openshift/installer/pkg/asset/tls.AdminKubeConfigSignerCertKey composite literal uses unkeyed fields                                                                                                                                                                       
pkg/install/kubeconfig_test.go:32:4: github.com/Azure/ARO-RP/vendor/github.com/openshift/installer/pkg/asset/tls.SelfSignedCertKey composite literal uses unkeyed fields  
```
not sure which key it complains about.

File: Makefile [hunk 10]
```diff
@@ -121,4 +121,7 @@ test-python: generate pyenv${PYTHON_VERSION}
 		azdev linter && \
 		azdev style
 
-.PHONY: aro az clean client generate image-aro proxy secrets secrets-update test-go test-python image-fluentbit publish-image-proxy publish-image-aro publish-image-fluentbit publish-image-proxy
+admin.kubeconfig:
+	hack/get-admin-kubeconfig.sh /subscriptions/${AZURE_SUBSCRIPTION_ID}/resourceGroups/${RESOURCEGROUP}/providers/Microsoft.RedHatOpenShift/openShiftClusters/${CLUSTER} >admin.kubeconfig
+
+.PHONY: aro az clean client generate image-aro proxy secrets secrets-update test-go test-python image-fluentbit publish-image-proxy publish-image-aro publish-image-fluentbit publish-image-proxy delete-cluster admin.kubeconfig
```
Comment (jim-minter) [inline Makefile]:
```suggestion
.PHONY: aro az clean client generate image-aro proxy secrets secrets-update test-go test-python image-fluentbit publish-image-proxy publish-image-aro publish-image-fluentbit publish-image-proxy admin.kubeconfig
```

File: pkg/util/restconfig/restconfig.go [hunk 11]
```diff
@@ -17,9 +17,12 @@ import (
 
 // RestConfig returns the Kubernetes *rest.Config for a kubeconfig
 func RestConfig(env env.Interface, oc *api.OpenShiftCluster) (*rest.Config, error) {
-	config, err := clientcmd.Load(oc.Properties.AdminKubeconfig)
+	config, err := clientcmd.Load(oc.Properties.AROServiceKubeconfig)
```
Comment (jim-minter) [inline pkg/util/restconfig/restconfig.go]: please decide which kubeconfig to use based on whether oc.Properties.AROServiceKubeconfig != "", then use that, rather than doing a fallback if clientcmd.Load fails for any reason

File: pkg/install/kubeconfig.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/ghodss/yaml"
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/tls"
+	clientcmd "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+// generateAROServiceKubeconfig generates additional admin credentials and kubeconfig
+// based on admin kubeconfig found in graph
+func (i *Installer) generateAROServiceKubeconfig(ctx context.Context, g graph, aroServiceName string) (*kubeconfig.AdminInternalClient, error) {
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: ctx argument is unused, please remove

File: pkg/install/kubeconfig.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+				Context: clientcmd.Context{
+					Cluster:  adminInternalClient.Config.Contexts[0].Context.Cluster,
+					AuthInfo: aroServiceName,
+				},
+			},
+		},
+		CurrentContext: aroServiceName,
+	}
+
+	data, err := yaml.Marshal(aroServiceInternalClient.Config)
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: Didn't we agree that generateAROServiceKubeconfig would return `[]byte, error`?

File: pkg/install/kubeconfig.go [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/openshift/installer/pkg/asset/tls"
+	clientcmd "k8s.io/client-go/tools/clientcmd/api/v1"
+)
+
+// generateAROServiceKubeconfig generates additional admin credentials and kubeconfig
+// based on admin kubeconfig found in graph
+func (i *Installer) generateAROServiceKubeconfig(ctx context.Context, g graph, aroServiceName string) (*kubeconfig.AdminInternalClient, error) {
+	ca := g[reflect.TypeOf(&tls.AdminKubeConfigSignerCertKey{})].(*tls.AdminKubeConfigSignerCertKey)
+	cfg := &tls.CertCfg{
+		Subject:      pkix.Name{CommonName: aroServiceName, Organization: []string{"system:masters"}},
```
Comment (jim-minter) [inline pkg/install/kubeconfig.go]: hard code aroService name to "system:aro-service" here

File: pkg/install/kubeconfig_test.go [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+				tls.CertKey{
+					CertRaw: pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: validCaCerts[0].Raw}),
+					KeyRaw:  pem.EncodeToMemory(&pem.Block{Type: "PRIVATE KEY", Bytes: b}),
+				},
+			},
+		},
+		reflect.TypeOf(&kubeconfig.AdminInternalClient{}): &kubeconfig.AdminInternalClient{},
+	}
+
+	APIServerUrl := "https://api.hash.rg.mydomain:6443"
```
Comment (jim-minter) [inline pkg/install/kubeconfig_test.go]: apiserverURL

File: pkg/install/kubeconfig_test.go [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	// validate the result in 2 stages: first verify the key and certificate
+	// are valid (signed by CA, have proper validity period, etc)
+	// then remove the AuthInfo struct from the result and validate
+	// rest of the fields by comparing with the template.
+
+	err = innercert[0].CheckSignatureFrom(validCaCerts[0])
+	if err != nil {
+		t.Error("Failed to parse certificate", err)
```
Comment (jim-minter) [inline pkg/install/kubeconfig_test.go]: t.Fatal(err)

File: pkg/install/kubeconfig_test.go [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	issuer := innercert[0].Issuer.String()
+	expectedIssuer := "CN=validca"
+	if issuer != expectedIssuer {
+		t.Errorf("Invalid certificate issuer, want: %s, got %s.", expectedIssuer, issuer)
```
Comment (jim-minter) [inline pkg/install/kubeconfig_test.go]: Just do `t.Error(issuer)` or equivalent for all of these rather than inventing error messages which are all differently formatted.

File: pkg/install/kubeconfig_test.go [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+
+	subject := innercert[0].Subject.String()
+	expectedSubject := "CN=system:aro-service,O=system:masters"
+	if subject != expectedSubject {
+		t.Errorf("Invalid subject want: %s, got %s.", expectedSubject, subject)
+	}
+
+	validity := innercert[0].NotAfter.Sub(innercert[0].NotBefore)
+	validityTenYears := time.Hour * 24 * 365 * 10
+	// truncate is needed because certificate validity can be slightly less or more than ten years
```
Comment (jim-minter) [inline pkg/install/kubeconfig_test.go]: something like `if !innercert[0].NotAfter.After(time.Now().AddDate(9, 11, 0)) { t.Error(...) }` ?

File: pkg/install/kubeconfig_test.go [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+	}
+
+	keyUsage := innercert[0].KeyUsage
+	expectedKeyUsage := x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature
+	if keyUsage != expectedKeyUsage {
+		t.Errorf("Invalid keyUsage.")
+	}
+
+	// validate the rest of the struct
+	got.AuthInfos = make([]clientcmd.NamedAuthInfo, 0, 0)
```
Comment (jim-minter) [inline pkg/install/kubeconfig_test.go]: `got.AuthInfos = []clientcmd.NamedAuthInfo{}`

File: pkg/install/kubeconfig_test.go (hunk 17 above)
Comment (olga-mir) [reply pkg/install/kubeconfig_test.go]: I like to have both values in the error message when error occur, it saves a lot of time for the developer to see it right in the pipeline logs, rather then go and reproduce on their own (and it might not reproduce in another environment) In my preference usability is more important in this case

File: pkg/install/kubeconfig.go (hunk 13 above)
Comment (olga-mir) [reply pkg/install/kubeconfig.go]: we have discussed this, but I am not convinced that this is necessarily better. In the way I suggest the function returns a complete structure, which aids with readability and usability. Especially that to make it a complete structure all it takes is just another text field. 
It also makes it clearer to people who read this code where this function is called:
```
		i.doc, err = i.db.PatchWithLease(ctx, i.doc.Key, func(doc *api.OpenShiftClusterDocument) error {
			// used for the SAS token with which the bootstrap node retrieves its
			// ignition payload
			doc.OpenShiftCluster.Properties.Install.Now = time.Now().UTC()
			doc.OpenShiftCluster.Properties.AdminKubeconfig = adminInternalClient.File.Data
			doc.OpenShiftCluster.Properties.AROServiceKubeconfig = aroServiceInternalClient.File.Data
			return nil
		})
```
in this way people can see that this is the same structure as the other one and they don't need to jump into the function to try to figure out what is that value that is returned.

**Comment by jim-minter** [comment]
rebased
